src/openvslam/data/map_database.cc

```diff
@@ -71,6 +71,69 @@ std::vector<keyframe*> map_database::get_all_keyframes() const {
     return keyframes;
 }
 
+std::vector<keyframe*> map_database::get_close_keyframes(const Mat44_t& pose,
+                                                         const unsigned int limit) const {
+    std::lock_guard<std::mutex> lock(mtx_map_access_);
+
+    // Auxilary vector of {keyframe_id, correlation_value} pair
+    // to have a key-frames sorted by their correlation with given pose
+    std::vector<std::pair<unsigned int, double>> correlations;
+
+    Vec3_t Ma = pose.block<3, 1>(0, 0);
+    Vec3_t Mt = pose.block<3, 1>(0, 3);
+    for (const auto id_keyframe : keyframes_) {
+        Vec3_t Na = id_keyframe.second->get_cam_pose().block<3, 1>(0, 0);
+        Vec3_t Nt = id_keyframe.second->get_cam_pose().block<3, 1>(0, 3);
+        // Angle between two cameras related to given pose and selected keyframe.
+        //
+        // Some background: there are two rotation matrixes: M and N,
+        // where M - is a pose rotation matrix, N - keyframe rotation matrix.
+        // Let's suppose we have a vector "a" in a coordinate system.
+        // After two rotations will be applied, "a" will be transformd to:
+        // b = M * a
+        // c = N * a
+        // in the same coordinate system.
+        //
+        // Angle between two vectors "b" and "c" is being calculated as follows:
+        // cos(angle) = (b, c) / ( len(b) * len(c) ).
+        // Since all "a", "b", "c" vectors have 1-length, the formula will be as follows:
+        // cos(angle) = (M * a, N * a)
+        //
+        // In OpenVSLAM camera direction (camera's normal) coincides with the OX axis.
+        // So, a = (1,0,0).
+        // Therefore, we have:
+        // cos(angle) = ( (M[0,0],M[1,0],M[2,0]) * (N[0,0],N[1,0],N[2,0]) )
+        double cos_angle = Ma.transpose() * Na;
```

Comment:
Do feature points in relocalization algorithms are resistant to angle-of-list? In other words, do we need to treat two frames looking in the same direction but tilted/heeling between each other as similar or not?

Reply:
I think it is possible to ignore the rotation around the optical axis.
I noticed that if we use Equirectangular, we can ignore the yaw angle.

Reply:
I am talking about roll angle, not yaw. If relocalization is not resistant to banks, we have to use `M * N.transpose()` angle. If does, this should be the angle between two normal vectors to cameras (current pose and estimated key-frame). That is the thing I am doubt about.

Reply:
I think `M * N.transpose()` is the better choice, but I think it might be a good idea to experiment and decide.

Reply:
Today I've made some experiments with the phone monocular camera (camera was preliminarily calibrated using OpenCV calibration [tutorial](https://docs.opencv.org/3.4/d4/d94/tutorial_camera_calibration.html)). The mapping dataset was recorded in the place full of features with zero bank angle. On this dataset I've build an OpenVSLAM map. Then I've started localization mode with this map. The tracking is not being lost only in the case when the phone is almost parallel to the ground level (0-bank angle). If turn the phone a little, the localization immediately disappears. After returning back it appears again:
![bank_experiment](https://user-images.githubusercontent.com/60094858/118855096-1b960880-b8de-11eb-847e-6fbee5ac2f7e.gif)

I think this should be prove enough to consider that localization is not resistant to banks and we need to calculate `M * N.transpose()` matrix' angle.

Reply:
It seems to be related to the fact that check_orientation is set to true in the following:

https://github.com/OpenVSLAM-Community/openvslam/blob/69b2bfaaf032b888f9223f3709fa0926fa7f8837/src/openvslam/match/bow_tree.h#L18
https://github.com/OpenVSLAM-Community/openvslam/blob/69b2bfaaf032b888f9223f3709fa0926fa7f8837/src/openvslam/module/relocalizer.cc#L18

Reply:
Done in https://github.com/OpenVSLAM-Community/openvslam/pull/79/files#diff-df221efd4ba8c25e4d8db053b418e43198ce2b6099e0c54b68ef139882692efeR94

Reply:
Why are you calculating cos_angle instead of angle?

Reply:
The main reason here is a performance. Cosine of angle calculated by using `(M * N.transpose()).trace()` is ~4.5 times faster than `Eigen::AngleAxisd(M * N.transpose()).angle()`. Measured on 192 angles, average time of first approach was about 25ns per one calculation comparing to 115ns of second approach.

Also, all angles are better sorted by similarity by using cosine of angle instead of absolute angle. For example `0++` and `2*pi--` angles are in the same similarity, but there is no simple way to compare them, without not using cosine. The meaning of cos(angle) - is a projection of turned coordinates vector into initial one which also fits within the meaning to term "similarity". And since the final similarity score depends on position in a sorted array rather than absolute value, cos(angle) will fit there, I suppose.